### PR TITLE
PyTorch & Tensorflow Interop v2: abstractions

### PR DESCRIPTION
**Pull request by alexcoca**
This PR follows on from #266. Changes include:

- The `Counterfactual` wrapper class is removed. The user interacts directly with classes implementing the specific algorithm they want to use for counterfactual search (e.g., `WachterCounterfactual`)

- `WachterCounterfactual` is a public version of a private class that implements the actual algorithm. The role of the public class is to implement the `alibi` API (referred to in the docs as an API class). This segregation of duty allows us to implement batch explanations (see #270 for example). All classes implementing counterfactual search methods inherit from `CounterfactualBase` class and implement its interface. 

- To achieve interoperability, the `WachterCounterfactual` class defers PyTorch/TensorFlow specific functionality to a `backend` class, which is loaded and initialised automatically by `CounterfactualBase` class

- The above is possible by using the `register_backend` decorator in `alibi.explainers.backend`. The decorator is parameterized by the `consumer_class` argument, which should specify the name of the class where this backend is going to be loaded. Optionally, the `predictor_type` argument can be used to black-box variants of a backend. 

- The backend classes are framework specific. For TensorFlow they inherit from `TFCounterfactualOptimizer`. At a minimum, the sub-classes should:

     * Initialise the superclass with a loss specification
     * Implement the `update_state`, `initialise_variables`, `autograd_loss`  using the loss specification 

- The backend offers the `.step()` method that calling objects can use in order to run one optimization step.

Next steps:

- Evaluate the usefuleness of the porposed abstractions
- Propose enhancements/alternative solutions

### CODE REVIEW

File: alibi/explainers/backend/tensorflow/counterfactuals.py [hunk 1]
```diff
@@ -0,0 +1,830 @@
...
+    if np.any(feat_interval < 0.):
+        raise ValueError(
+            f"Invalid constrain. Lower bound needs to be smaller upper bound but the difference was "
+            f"{feat_interval} for an element.",
+        )
+
+    return tf.clip_by_value(X, clip_value_min=low, clip_value_max=high)
+
+
+def slice_prediction(prediction: tf.Tensor, target_idx: Union[int, str], src_idx: int) -> tf.Tensor:
```
Comment (jklaise) [inline alibi/explainers/backend/tensorflow/counterfactuals.py:184]: Copy-paste from previous PR:
>Can use Literal for types if it doesn't become too unreadable.

File: alibi/explainers/backend/tensorflow/counterfactuals.py [hunk 2]
```diff
@@ -0,0 +1,830 @@
...
+
+
+def wachter_blackbox_wrapper(X: Union[tf.Tensor, tf.Variable],
+                             predictor: Callable,
+                             target_idx: Union[str, int],
+                             src_idx: int) -> tf.Tensor:
+
+    """
+    A wrapper that modifies a predictor that returns a vector output when queried with input :math:`X` predictor that
+    returns a vector to return a scalar. The scalar depends on the `target_idx` and `src_idx`. The predictor (and the
```
Comment (jklaise) [inline alibi/explainers/backend/tensorflow/counterfactuals.py:225]: This first sentence need to be cleaned up for clarity.

File: alibi/explainers/backend/tensorflow/counterfactuals.py [hunk 3]
```diff
@@ -0,0 +1,830 @@
...
+        Updates the current solution with the gradients of the loss.
+
+        Parameters
+        ----------
+        gradients
+            A list containing the gradients of the differentiable part of the loss.
+        """
+
+        # TODO: ALEX: TBD: SHOULD SELF.SOLUTION BE A LIST? I
+        # TODO: ALEX: TBD: THIS MASK IS NOT TOO GENERIC? WE COULD IF ELSE ON IT?
```
Comment (jklaise) [inline alibi/explainers/backend/tensorflow/counterfactuals.py:442]: Let's discuss as I'm not sure what the issues are.

File: alibi/explainers/backend/tensorflow/counterfactuals.py [hunk 4]
```diff
@@ -0,0 +1,830 @@
...
+            loss_spec = WACHTER_LOSS_SPEC_WHITEBOX
+        super().__init__(predictor, loss_spec, feature_range, **kwargs)
+
+        # track the total loss
+        self.state.update([('total_loss', 0.0)])
+
+        # updated by the calling context in the optimisation loop
+        self.lam = None  # type: Union[float, None]
+
+        # TODO: ALEX: TBD: DESIGN ALTERNATIVES WRT TO WHAT PREDICTOR RETURNS
```
Comment (jklaise) [inline alibi/explainers/backend/tensorflow/counterfactuals.py:537]: Let's discuss as I'm not sure what the issues are.

File: alibi/explainers/backend/tensorflow/counterfactuals.py [hunk 5]
```diff
@@ -0,0 +1,830 @@
...
+        Adds additional attributes to the class to support numerical differentiation of black-box predictors:
+
+            - `num_grad_fcn`: this a numerical differentiation procedure that, given a callable, an input :math:`X`, a \
+            set of keyword arguments and a set of additional arguments and keyword arguments for the callable, evaluates \
+            the gradient of the callable with respect to the input :math:`X` 
+            - `prediction_grad_fcn`: the derivative of :math:`L_{pred}` with respect to `pred`, the predictor output
+        """  # noqa W605
+        super().__init__(predictor, loss_spec, feature_range, **kwargs)
+
+        # TODO: ALEX: TBD: THIS PRACTICALLY MEANS ATTRS NOT IN THE SCHEMA WILL NOT BE SET
```
Comment (jklaise) [inline alibi/explainers/backend/tensorflow/counterfactuals.py:698]: Sounds reasonable to me?

File: alibi/explainers/backend/common.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
+# flake8: noqa: F401
```
[the remark this answers is not in the document]
Comment (alexcoca) [reply alibi/explainers/backend/common.py:1]: No, other than "I forgot".

File: alibi/explainers/cf_example_bb.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
+import tensorflow as tf
```
[the remark this answers is not in the document]
Comment (alexcoca) [reply alibi/explainers/cf_example_bb.py:1]: This script was there so you can run the code, it won't be part of any merge.

File: alibi/explainers/cf_example_wb.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
+import tensorflow as tf
```
[the remark this answers is not in the document]
Comment (alexcoca) [reply alibi/explainers/cf_example_wb.py:1]: See my previous comment.

File: alibi/utils/frameworks.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+def tensorflow_installed() -> bool:
+    """
+    Raise an ImportError if TensorFlow is not installed.
+    If TensorFlow>=2.0.0 is installed, issue a warning that some functionality may not work.
+    If TensorFlow<2.0.0 is installed, issue a warning that in the future some functionality will require an upgrade.
+    """
+
+    template = "Module requires {pkg}: pip install alibi[tensorflow]"
+    # TODO: ALEX: TBD: This should return False, and the error should be raised in files that req. TF.
```
[the remark this answers is not in the document]
Comment (alexcoca) [reply alibi/utils/frameworks.py:33]: This becomes a non-problem if we merge the TF2.0 upgrade proposal?

File: alibi/utils/gradients.py [hunk 10]
```diff
@@ -1,6 +1,26 @@
...
+"""
+dict: A registry for numerical gradient functions implemented in PyTorch or TensorFlow. This should be imported by
+backend classes that need to utilize numerical differentiation. To add a function to the registry, import the
+`numerical_gradient` decorator from `alibi.utils.gradients` and parametrize it by with the `framework` argument that
+takes values 'pytorch'/'tensorflow'.
+"""
+
+
+def numerical_gradient(framework='tensorflow'):
+    """
```
[the remark this answers is not in the document]
Comment (alexcoca) [reply alibi/utils/gradients.py:16]: I guess we can, I added the validation to make it easy to understand if you have typos, primarily.

File: alibi/utils/logging.py [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+        data_store
+            A dictionary with the structure::
+
+                {
+                    'var_name': [value],
+                }
+
+            It should contain all the variables to be logged. For a variable to be logged, it has to be defined in
+            the ``'tracked_variables'`` attribute of the default logging configuration that initialises this object.
+            If you don't want to update the logging configuration, use `record_step` instead.
```
[the remark this answers is not in the document]
Comment (alexcoca) [reply alibi/utils/logging.py:184]: Yes, that's correct. I thought the main mechanism should be `record`, which is why I only specified that one in the base class.

File: alibi/utils/tensorflow/logging.py [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+
+        with tf.summary.record_if(tf.equal(step % self.summary_freq, 0)):
+            with self.writer.as_default():
+                for variable in self.tracked_variables:
+                    if variable not in data_store:
+                        continue
+                    tag, var_type, description = self._vars_mapping[variable]
+                    if var_type == 'image' and step % self.image_summary_freq > 0:
+                        continue
+                    prefixed_tag = os.path.join(prefix, tag)
```
[the remark this answers is not in the document]
Comment (alexcoca) [reply alibi/utils/tensorflow/logging.py:68]: yes. Just a pref to use `os.path.join` vs `f"{prefix}/{tag}"`

File: alibi/explainers/experimental/counterfactuals.py [hunk 13]
```diff
@@ -0,0 +1,1264 @@
...
+    'Current lambda value',
+    'Lambda lower bound',
+    'Lambda upper bound',
+    'Difference between upper and lower bound',
+    '',
+    '',
+    'The probability output by the model for the current solution for the class of the instance whose cf is searched. ',
+    '',
+    ''
+]
```
[the remark this answers is not in the document]
Comment (alexcoca) [reply alibi/explainers/experimental/counterfactuals.py:74]: I thought the explanations for those variables were mostly trivial? Also thought I'd make it obvious these can be omitted!

File: alibi/explainers/experimental/counterfactuals.py [hunk 14]
```diff
@@ -0,0 +1,1264 @@
...
+            raise CounterfactualError(
+                "For non-differentiable predictors, a 'gradient_method' must be specified in the loss specification. "
+                "Available numerical differentiation methods are 'num_grad_batch'."
+            )
+        else:
+            if grad_method_name not in available_num_grad_fcns:
+                raise CounterfactualError(
+                    f"Undefined numerical gradients calculation method. Avaialble methods are "
+                    f"{available_num_grad_fcns}"
+                )
```
[the remark this answers is not in the document]
Comment (alexcoca) [reply alibi/explainers/experimental/counterfactuals.py:251]: Simplified this a bit.

File: alibi/explainers/base/counterfactuals.py [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+        self.step = -1
+        self.lam_step = -1
+
+    def setup_tensorboard(self):
+        """
+        Initialises the TensorBoard writer. This method should be called at the beginning of the `counterfactual` method.
+        """  # noqa
+
+        if self.log_traces:
+            self.tensorboard = self.tensorboard().setup(self.logging_opts)
```
[the remark this answers is not in the document]
Comment (alexcoca) [reply alibi/explainers/base/counterfactuals.py:202]: I don't know. I can make the object itself a private attribute, but the idea here is to only initialise the writer when all the options are known.

File: alibi/explainers/experimental/counterfactuals.py [hunk 16]
```diff
@@ -0,0 +1,1264 @@
...
+            tensor) should always be two-dimensional. For example, a single-output classification model operating on a 
+            single input instance should return a tensor or array with shape  `(1, 1)`. The explainer assumes the 
+            `predictor` returns probabilities. In the future this explainer may be extended to work with regression 
+            models. If `predictor_type` is set to `blackbox` then the `predictor` should be a callable that inputs/outputs
+            `np.ndarray` objects.
+         predictor_type: {'blackbox', 'whitebox'}
+
+            - 'blackbox' indicates that the algorithm does not have access to the model parameters (e.g., the predictor \
+            is an API endpoint so can only be queried to return prediction). This argument should be used to search for \
+            counterfactuals of non-differentiable models (e.g., trees, random forests) or models implemented in machine \
```
[the remark this answers is not in the document]
Comment (alexcoca) [reply alibi/explainers/experimental/counterfactuals.py:1023]: I already stated I disagree with the previous statement.

File: alibi/explainers/experimental/counterfactuals.py [hunk 17]
```diff
@@ -0,0 +1,1264 @@
...
+
+class WachterCounterfactual(Explainer, FitMixin):
+
+    # TODO: ALEX: TBD: AS WE IMPLEMENT COUNTERFACTUALS, WE MAY REALISE THIS IS COMMON AND WOULD DO THE FOLLOWING:
+    #  - make this a superclass (e.g., CounterfactualAPI). The public classes (e.g., WachterCounterfactual) inherit from
+    #  it to show the docs specific to the method and potentially override some behaviour (eg. fit might have other
+    #  set of arguments, etc)
+    def __init__(self,
+                 predictor: Union[Callable, 'tf.keras.Model', 'keras.Model'],
+                 predictor_type: Literal['blackbox', 'whitebox'] = 'blackbox',
```
[the remark this answers is not in the document]
Comment (alexcoca) [reply alibi/explainers/experimental/counterfactuals.py:997]: I don't like the idea of "inferring" because it is usually brittle and does not generalise well. The users of the library should specify what use case they have.

File: alibi/explainers/cf_example_bb.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (alexcoca) [reply alibi/explainers/cf_example_bb.py:1]: These were added in case you wanted to run the code, they're going to be removed in a final version.

File: alibi/utils/frameworks.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (jklaise) [reply alibi/utils/frameworks.py:35]: Not exactly because in the future TF will be optional, so we need to check if the user has TF installed but also which version.

File: alibi/utils/tensorflow/logging.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (jklaise) [reply alibi/utils/tensorflow/logging.py]: Imo we should stick with strings as it's semantically more clear.

**Comment by alexcoca** [comment]
Closing - to be merged into dev branch in alibi.
